tensorflow/python/keras/wrappers/scikit_learn.py

```diff
@@ -71,285 +92,663 @@ class BaseWrapper(object):
+
+        Arguments:
+            deep : bool, default=True
+                If True, will return the parameters for this estimator and
+                contained subobjects that are estimators.
+
+        Returns:
+            params : mapping of string to any
+                Parameter names mapped to their values.
+        """
+        out = dict()
+        for key in self.init_params:
+            value = getattr(self, key)
+            if deep and hasattr(value, "get_params"):
+                deep_items = value.get_params().items()
+                out.update((key + "__" + k, val) for k, val in deep_items)
+            out[key] = value
+        return out
+
+    def set_params(self, **params):
+        """
+        Set the parameters of this estimator.
+        The method works on simple estimators as well as on nested objects
+        (such as pipelines). The latter have parameters of the form
+        ``<component>__<parameter>`` so that it's possible to update each
+        component of a nested object.
+
+        This method mimics sklearn.base.BaseEstimator.set_params
+        
+        Arguments:
+            **params : dict
+                Estimator parameters.
+        Returns:
+            self : object
+                Estimator instance.
+        """
+        if not params:
+            # Simple optimization to gain speed
+            return self
+        valid_params = self.get_params(deep=True)
+
+        nested_params = defaultdict(dict)  # grouped by prefix
+        for key, value in params.items():
+            key, delim, sub_key = key.partition("__")
+            if key not in valid_params:
+                raise ValueError(
+                    "Invalid parameter %s for estimator %s. "
+                    "Check the list of available parameters "
+                    "with `estimator.get_params().keys()`." % (key, self)
+                )
+            if delim:
+                nested_params[key][sub_key] = value
+            else:
+                setattr(self, key, value)
+                valid_params[key] = value
+
+        for key, sub_params in nested_params.items():
+            valid_params[key].set_params(**sub_params)
+
+        return self
+
+    def __getstate__(self):
+        """Used by various scikit-learn methods to clone estimators. Also used
+           for pickling.
+           Because some objects (mainly Keras `Model` instances) are not pickleable,
+           it is necessary to iterate through all attributes and clone the
+           unpicklables manually.
+
+        Returns:
+            state : dictionary containing a copy of all attributes of this 
+                    estimator with Keras Model instances being saved as 
+                    HDF5 binary objects.
+        """
+
+        def __pack_obj(obj):
+            if hasattr(obj, "save"):  # for models
+                as_bytes = io.BytesIO()
+                with h5py.File(as_bytes, mode="w") as file:
+                    obj.save(file)
```

Comment:
This should be doable without `h5py`. I'm not sure what the prefered format to use is, or if it is better to use `serialize_keras_object` or something like that.

Reply:
I take this back: the entirety of Tensorflow/Keras is hardcoded to filenames or `h5py.File` objects. 
 
An aside, unrelated to this PR: It is a real shame that we can't save to arbitrary file streams. I understand that HDF5 can have advantages over a more straightforward compress and dump to file stream method, but if it is the _only_ way to save to an io stream, it should be made a requirement, not a sort of optional dependancy. Maybe I'll submit a seperate PR to enable at least `io.BytesIO`.

Reply:
@qlzh727 I know you said in the other PR that you don't really work with the sklearn wrappers, but any input on this more general issue of how to serialize Keras models to memory in a picklable form?

Reply:
I came up with the following. It _seems_ to work for all of the types of models I had, but maybe it fails for custom models?
```python3
from tensorflow.python.keras.saving import saving_utils


SavedKerasModel = namedtuple(
    "SavedKerasModel", "cls model_config training_config weights",
)
```


```python3
def __pack_obj(obj):
    try:
        return copy.deepcopy(obj)
    except TypeError:
        if isinstance(obj, Model):
            model_metadata = saving_utils.model_metadata(obj)
            training_config = model_metadata["training_config"]
            model_config = obj.get_config()
            weights = obj.get_weights()
            return SavedKerasModel(
                cls=obj.__class__,
                model_config=model_config,
                weights=weights,
                training_config=training_config,
            )
```
```python3
def __unpack_obj(obj):
    if isinstance(obj, SavedKerasModel):
        restored_model = obj.cls.from_config(obj.model_config)
        training_config = obj.training_config
        restored_model.compile(
            **saving_utils.compile_args_from_training_config(training_config)
        )
        restored_model.set_weights(obj.weights)
        return restored_model
```

Reply:
@k-w-w who is the expert for save model.

Reply:
Any updates @qlzh727 @k-w-w ?

Reply:
This wrapper looks pretty good and should work for non-subclassed models. To support subclassed models, you should save both the class name and config (the results from [generic_utils.serialize_keras_object](https://github.com/tensorflow/tensorflow/blob/v2.1.0/tensorflow/python/keras/utils/generic_utils.py#L191)). 

Calling [`keras.layers.deserialize`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/layers/serialization.py#L71) will load subclassed models if they were registered with [`tf.keras.utils.generic_utils.register_keras_serializable`](https://www.tensorflow.org/api_docs/python/tf/keras/utils/register_keras_serializable).

Reply:
Thank you @k-w-w. I will look into implementing that. Any chance you can do a quick approval, just so all of the tests can run?